app_dart/lib/src/request_handlers/update_agent_health_history.dart

```diff
@@ -0,0 +1,92 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:collection/collection.dart';
+import 'package:gcloud/db.dart';
+import 'package:googleapis/bigquery/v2.dart';
+import 'package:meta/meta.dart';
+
+import '../datastore/cocoon_config.dart';
+import '../model/appengine/agent.dart';
+import '../request_handling/api_request_handler.dart';
+import '../request_handling/authentication.dart';
+import '../request_handling/body.dart';
+import '../service/datastore.dart';
+
+const Duration maxHealthCheckAge = Duration(minutes: 10);
+
+@immutable
+class UpdateAgentHealthHistory extends ApiRequestHandler<Body> {
+  const UpdateAgentHealthHistory(
+    Config config,
+    AuthenticationProvider authenticationProvider, {
+    @visibleForTesting
+        this.datastoreProvider = DatastoreService.defaultProvider,
+  }) : super(config: config, authenticationProvider: authenticationProvider);
+
+  final DatastoreServiceProvider datastoreProvider;
+
+  @override
+  Future<Body> get() async {
+    const String projectId = 'flutter-dashboard';
+    const String dataset = 'cocoon';
+    const String table = 'Agent';
+
+    final TabledataResourceApi tabledataResourceApi =
+        await config.createTabledataResourceApi();
+    final DatastoreService datastore = datastoreProvider();
+    final Query<Agent> agentQuery = datastore.db.query<Agent>()
+      ..order('agentId');
+    final List<Agent> agents =
+        await agentQuery.run().where(_isVisible).toList();
+    agents.sort((Agent a, Agent b) =>
+        compareAsciiLowerCaseNatural(a.agentId, b.agentId));
+
+    final List<Map<String, Object>> tableDataInsertAllRequestRows =
+        <Map<String, Object>>[];
+
+    for (Agent agent in agents) {
+      final bool isHealthy = _isAgentHealthy(agent);
+
+      /// Consolidate [agents] together.
+      ///
+      /// Prepare for bigquery [insertAll].
+      tableDataInsertAllRequestRows.add(<String, Object>{
+        'json': <String, Object>{
```

Comment:
Is it possible to update `Agent.toJson()` to this instead?

Reply:
Agent model have more other properties. Here these four properties are in accordance with those defined in BigQuery table.

Reply:
It seems like the only thing not in accordance is the `capabilities` field. If BigQuery does not support arrays, couldn't we still just do `agent.toJson()..remove('capabilities')` to make this more concise?

Reply:
Yeah, that would make it more concise for sure.
However, another change is from bool `IsHealthy` to String `Status`.
Is it easy to do type cast here?

Reply:
In that case, this is probably fine. Is there a reason for switching it to type String?

Reply:
We may add more detailed categories, like maintenance, dead, quarantined, etc.

Reply:
Oh okay, then this is probably best. Might be good to add a TODO to remind us why it was done this way.

Reply:
Will add that when reverting earlier changes. Here is the TODO in my earlier PR:
https://github.com/flutter/cocoon/blob/master/app_dart/lib/src/request_handlers/update_agent_health.dart#L75